### DISCUSSION TITLE
Storage devices and the Virtual Chassis model

### DISCUSSION THREAD

**Discussion by mgoetze5** [netbox-community/netbox#5789] [General]
So I have a bunch of storage devices such as HPE 3PARs, NetApp FAS/AFF, EMC VNX which consist of multiple chassis, some of which have controller nodes and disks, others which have only disks. Each chassis has it's own asset number in our asset DB, and they could theoretically be non-contiguous (even though this is usually not best practice). So I have to make each chassis its own device in NetBox, however they are not independent, from a logical point of view they act as one device.
After scouring the NetBox documentation, it seems to me that this is supposed to be modelled as a Virtual Chassis in NetBox, correct?
I have a few issues with this, relating to how the Virtual Chassis model is implemented in NetBox. Mainly, while stacked switches may have a "master", there is no such thing for most storage devices. If I assign a master at random, I end up with inconsistent naming, e.g.

* 3par01 cage0
* 3par01 cage1
* 3par01.example.com
* 3par01 cage3

In the example above I have a 3PAR with cage2 chosen as the master. Not only is the naming inconsistent, but it seems that all the interfaces now appear under 3par01.example.com, which is great for the storage admin who wants to have a logical look at the device, but perhaps somewhat confusing for the datacenter technician who wants to see which cables physically belong to cage2.
As such I am considering a workaround whereby I add a 0U virtual device to the virtual chassis to act as the master. I wish I didn't have to assign it a position, as it has no cage or shelf ID, but I guess I can give it 99 or something. I noticed that it doesn't have an interfaces tab unless I give it an interface, but I suppose in most cases I want to give it a virtual mgmt interface anyway.
Thoughts on this idea? How do you manage multi-chassis storage systems in NetBox?

Edited to add: I have just tried adding individual controller nodes via device bays. Their interfaces don't show up in the virtual chassis unless I add the controller nodes (not just the chassis) to the virtual chassis. Again, it would be nice if I could do this without assigning a position, as they don't have their own cage/shelf ID.

**Comment by jhammond-git** [comment]
When you add interfaces to a device that's in a device bay, they don't show up on the parent.  You still have to click through to the child to see the interfaces.  So that makes sense that you'd have to add the child devices directly to the VC to get the interfaces to show up.

I would think that the Interfaces tab not showing up on a master device just because it doesn't directly have interfaces could be a bug, but I don't have access to an older version to see what how it used to work before the changes to VCs.

For our storage systems, we don't try to group everything together.  The shelves and controllers are all grouped via naming convention.  It may be that consistency is the enemy of correct in your case.

**Reply by mgoetze5** [reply]
Thanks for your input. Would you mind giving an example from your setup? In the end you do have to assign a primary management IP to something (which ideally has the fqdn of the whole storage device), don't you?

**Reply by jhammond-git** [reply]
In our case, no storage array crosses racks, so my example may not be valid for you.  Having said that, for a NetApp we have:

Given a <system-name>
<system-name>-01 = controller A
<system-name>-02 = controller B
<system-name> - ID 1.01 = Standard storage shelf 1
<system-name> - ID 1.02 = Standard storage shelf 2
.
.
.
etc...

None of them are part of a virtual chassis.  Each device is separate and racked individually in Netbox.

Looking at your example again, I'm wondering if creating a VC with no master might be a better option for you.  In more recent versions of Netbox, you can do so.  The difference being, no device displays all the interfaces on the chassis - each member only displays their own interfaces.

**Reply by mgoetze5** [reply]
But then where do you register your cluster_mgmt LIF? Or are you not using NetBox for IPAM?
I guess I get the feeling that you're using NetBox for different purposes than we're planning to. (Still interesting to see how other people are doing it, though!)

**Reply by jhammond-git** [reply]
We do use it for IPAM, however, you're getting into areas that I don't know about.  A lot of the storage bits, once you get beyond the physical units, is a mystery to me.  I had to look up the term cluster_mgmt LIF, and it seems I have a question for our storage engineers to clear up about where, if anywhere, that's documented.

**Reply by airborneric** [reply]
Has anyone been able to figure this out yet?  Having a similar issue where we are primarily using 2 node clusters, so netbox works okay for these.  We cannot use netbox for more than 2 node clusters (ha pairs) as the cluster IP is "locked" to the primary node e0M port.  I can make it an anycast, but it is still "locked" on node-01 of the original HA pair.

I am going to explore the Virtual Cluster piece, but currently I don't see a good way to model storage inside Netbox and make it technically accurate for a "source of truth".

**Reply by airborneric** [reply]
Actually, I do not know if I am allowed to create a Virtual Cluster (VC)? When I add an A800, it locks the chassis into a rack unit - then the nodes are considered device bays.  This in turn locks the cluster in whatever rack it is assigned?  Cannot extend past that chassis.  Technically, the NetApp cluster can consist of 2-24 nodes (ha pairs) and span multiple racks.  I do not see a way to do this inside Netbox.....

**Reply by mgoetze5** [reply]
I ended up going with the virtual master device approach, so for NetApp specifically I have a device type "ONTAP Cluster" with no height that I use as the master of the Virtual Chassis. With regard to multirack, when creating a Virtual Chassis if you specify a rack you can only initially select devices from that rack, but you can still later add devices from any rack, also technically you don't have to assign the Virtual Chassis to a rack.
So for a medium-sized cluster I might have a Virtual Chassis with the following devices in it (device types in parentheses):

napp01 (ONTAP Cluster) - Master device
napp01 Chassis-01 (A700 Chassis)
napp01-01 (A700 Controller Node)
napp01-02 (A700 Controller Node)
napp01 Chassis-01 Shelf-00 (DS224C)
napp01 Chassis-01 Shelf-10 (DS224C)
napp01 Chassis-02 Shelf-00 (A220 Chassis)
napp01-03 (A220 Controller Node)
napp01-04 (A220 Controller Node)
napp01 Chassis-02 Shelf-01 (DS224C)

**Reply by airborneric** [reply]
Will see if this works for us - thank you!

**Reply by airborneric** [reply]
Sorry - where did you put the Cluster-mgmt IP?  Can you put that on the master device?

**Reply by mgoetze5** [reply]
Yes, the master device has a virtual interface called cluster_mgmt and the IP assigned there is also the primary IP of the master device.
Other than that to be honest I'm somewhat undecided whether to model every LIF as its own virtual interface, currently we just assign all the LIF IPs other than cluster_mgmt to their home port.

**Reply by airborneric** [reply]
Thank you!  Will give this a go!
